### ISSUE TITLE
It seems to be impossible to exclude parsing the entire typescript package source code when including it

### ISSUE THREAD

**Issue by ChristianIvicevic** [typescript-eslint/typescript-eslint#2038]
**Repro**

Using the following `.eslintrc.json` in a generic CRA

```JSON
{
    "extends": [
        "react-app",
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:import/errors",
        "plugin:import/warnings",
        "plugin:import/typescript",
        "plugin:jest/recommended",
        "prettier/@typescript-eslint",
        "plugin:prettier/recommended"
    ],
    "parser": "@typescript-eslint/parser",
    "env": { "node": true },
    "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module",
        "project": "./tsconfig.eslint.json",
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "overrides": [
        {
            "files": ["**/*.ts?(x)"],
            "rules": {
                // ...
            }
        }
    ]
}
```

together with the following two `tsconfig` files

```JSON
{
    "compilerOptions": {
        "target": "es5",
        "lib": ["dom", "dom.iterable", "esnext", "esnext.intl", "es2017.intl", "es2018.intl"],
        "allowJs": true,
        "skipLibCheck": true,
        "esModuleInterop": true,
        "allowSyntheticDefaultImports": true,
        "strict": true,
        "forceConsistentCasingInFileNames": true,
        "module": "esnext",
        "moduleResolution": "node",
        "resolveJsonModule": true,
        "isolatedModules": true,
        "downlevelIteration": true,
        "noEmit": true,
        "jsx": "react",
        "baseUrl": "src"
    },
    "include": ["src"],
    "exclude": ["node_modules"]
}
```

```JSON
{
    "extends": "./tsconfig.json",
    "include": ["src", "*.config.js"],
    "exclude": ["node_modules"]
}
```

and an `.eslintignore`

```
node_modules
```

the following code opened in VSCode (or being linted via the command line)

```TS
import ts from "typescript";

// eslint-disable-next-line @typescript-eslint/no-floating-promises
void (async function () {
    // few lines of code that have some linter errors
})();
```

**Expected Result**

Linting the file in a few ms and fixing issues in a few ms.

**Actual Result**

The above code takes about to 30-60s to lint. In VSCode especially I notice that the warnings and errors take about 20s to load in, and fixing linter errors takes 60s with a VSCode popup appearing indicating the unresponsiveness of the task.

![image](https://user-images.githubusercontent.com/45569997/82233469-5b738a00-9930-11ea-833f-1baa7d8288a4.png)

My arbitrary guess is that the linter tries to include the entirety of the typescript definitions into the linting process for improved results of rules that require type information. Furthermore it seems no steps to exclude the typescript package worked, but I don't want to opt out of linting of the script file. I just want to let eslint ignore the types of the typescript package.

Linting the entire project (apart from this particular file) takes about 8s which is quick enough, with that single script file it basically increases this time by approximately x6-8.

**Additional Info**

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.33.0` |
| `TypeScript`                | `3.9.2` |
| `ESLint`                    | `6.8.0` |
| `node`                      | `12.16.1` |

**Comment by bradzacher** [comment]
> I just want to let eslint ignore the types of the typescript package.

That's not possible.

> My arbitrary guess is that the linter tries to include the entirety of the typescript definitions

The parser is built on top of typescript. All we do is use TS's compiler APIs to parse files (and generate types if required), then convert typescript's AST to one understandable via ESLint.

When you import a module, it is TS that resolves that import and all of its dependencies.
Thus, TS controls the file loading, type resolution and parsing.
Just like you cannot exclude the type declarations for the `typescript` module from a `tsc` run, you cannot exclude them from the lint run.

If this were possible, this would also be inadvisable as this would lead to the import for the module being implicitly typed as `any`, which means our lint rules would detect everything as being typed as `any`, and thus would either false negative, or false positive.

----

Could you create a repro repo? There's likely another reason that it's taking so long, but it's impossible to investigate this as is.

**Comment by ChristianIvicevic** [comment]
Thank you for explaining the internals. I'll try to create a smaller repro repo once I find some time and will update you here once I'm done.
